File: code/SoundPlayground/SoundPlayground/Core/SharedContext.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SoundPlayground.Core {
    public class SharedContext {
        public int ChannelCount { get; set; } = 24;

        public Dictionary<int, Instrument> Channels { get; set; } = new Dictionary<int, Instrument>();

        public IEnumerable<int> AvailableChannels {
            get {
                return Enumerable.Range( 0, ChannelCount ).Where( c => !Channels.ContainsKey( c ) );
            }
        }

        public void RegisterInstrument ( Instrument instrument ) {
            int? channel = AvailableChannels.Select( a => (int?)a ).FirstOrDefault();

            if ( channel == null ) {
                throw new System.Exception( "No channel available found." );
            }

            Channels[ channel.Value ] = instrument;

            instrument.Channel = channel.Value;
        }

        public void UnregisterInstrument ( Instrument instrument ) {
            Channels.Remove( instrument.Channel.Value );

            instrument.Channel = null;
        }
    }
}

File: code/SoundPlaygroundPy/core/shared_context.py
from py_linq import Enumerable

class SharedContext():
    def __init__ ( self ):
        self.channel_count = 24
        self.channels = dict()

    @property
    def available_channels ( self ):
        return Enumerable.range( 1, self.channel_count )\
            .where( lambda i: i not in self.channels )

    def register_instrument ( self, instrument ):
        channel = self.available_channels\
            .first_or_default()

        if channel == None:
            raise BaseException( "No channel available found" )

        self.channels[ channel ] = instrument

        instrument.channel = channel

    def unregister_instrument ( self, instrument ):
        del self.channels[ instrument.channel ]

        instrument.channel = None
